misc/multistream-select/src/length_delimited.rs

```diff
@@ -323,27 +323,11 @@ where
     R: AsyncWrite
 {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        // Try to drain the write buffer together with writing `buf`.
```

Comment:
So that was the issue with the corrupted data?

Reply:
I think so. I find this code difficult to understand but here is an example where I think it fails:

- let `|write_buffer|` = 1 and `|buf|` = 2;
- let `poll_write_buffer` write all 3 bytes, leaving an empty `write_buffer`.
- `written` would be calculated as 1 - 0 = 1.
- `written` < 2 holds, and so `Ok(1)` is returned where in reality all 2 bytes from `buf` have been written.

This leads to data duplication which causes subsequent parsing errors on the receiving side.

Another problem is that we might run into arithmetic underflow when calculating `written` if less than `|buf|` bytes are written by `poll_write_buffer` and consequently the length of `write_buffer` is still larger than when calculating `n`.